### PR TITLE
TS Strict Stately List

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

File: packages/@react-stately/list/src/useListState.ts [hunk 1]
```diff
@@ -57,44 +60,48 @@ export function useListState<T extends object>(props: ListProps<T>): ListState<T
   );
 
   // Reset focused key if that item is deleted from the collection.
-  const cachedCollection = useRef(null);
+  const cachedCollection = useRef< Collection<Node<T>> | null>(null);
```
Comment (ktabors) [inline packages/@react-stately/list/src/useListState.ts:63]:
```suggestion
  const cachedCollection = useRef<Collection<Node<T>> | null>(null);
```

File: packages/@react-stately/list/src/useListState.ts [hunk 2]
```diff
@@ -57,44 +60,48 @@ export function useListState<T extends object>(props: ListProps<T>): ListState<T
...
         (
           diff > 1 ?
-          Math.max(startItem.index - diff + 1, 0) :
-          startItem.index
+          Math.max((startItem?.index ?? 0) - diff + 1, 0) :
+          startItem?.index ?? 0
         ),
         itemNodes.length - 1);
-      let newNode:Node<T>;
+      let newNode:Node<T> | null = null;
```
Comment (ktabors) [inline packages/@react-stately/list/src/useListState.ts:87]:
```suggestion
      let newNode: Node<T> | null = null;
```

**Review by ktabors — APPROVED** [review]

File: packages/@react-spectrum/tabs/src/Tabs.tsx [hunk 3]
```diff
@@ -335,6 +339,8 @@ export function TabPanels<T>(props: SpectrumTabPanelsProps<T>) {
   const {tabListState} = tabState;
 
   const factory = useCallback(nodes => new ListCollection(nodes), []);
+  // TODO: how to fix this one?
```
Comment (yihuiliao) [inline packages/@react-spectrum/tabs/src/Tabs.tsx:342]: what was the issue here?

Comment (snowystinger) [reply packages/@react-spectrum/tabs/src/Tabs.tsx:349]: added comment with the error i haven't been able to resolve yet

**Review by yihuiliao — APPROVED** [review]
i know there's still that one ts error with useCollection but the rest looks good. i tried poking around a little bit but couldn't figure it out either

File: packages/@react-spectrum/tabs/src/Tabs.tsx (hunk 3 above)
Comment (LFDanLu) [reply packages/@react-spectrum/tabs/src/Tabs.tsx:349]: Do we need to extend the TabPanel generic from object like we do for our other collection components? Then perhaps nodes needs to be defined as `(nodes: Iterable<Node<T>>)` in the factory callback?

**Review by LFDanLu — COMMENTED** [review]
looks fine to me, added as suggestion for the tabpanel one

**Review by yihuiliao — DISMISSED** [review]

File: packages/@react-spectrum/tabs/src/Tabs.tsx [hunk 4]
```diff
@@ -392,11 +405,11 @@ function TabPicker<T>(props: TabPickerProps<T>) {
     visible
   } = props;
 
-  let ref = useRef(undefined);
-  let [pickerNode, setPickerNode] = useState(null);
+  let ref = useRef(null) as DOMRef<HTMLDivElement> | undefined;
```
Comment (devongovett) [inline packages/@react-spectrum/tabs/src/Tabs.tsx:408]: you shouldn't need these casts

File: packages/@react-spectrum/tabs/src/Tabs.tsx (hunk 3 above)
Comment (devongovett) [reply packages/@react-spectrum/tabs/src/Tabs.tsx:349]: I pushed a fix

File: packages/@react-stately/combobox/src/useComboBoxState.ts [hunk 5]
```diff
@@ -176,13 +190,15 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateOptions<T
 
   let [lastValue, setLastValue] = useState(inputValue);
   let resetInputValue = () => {
-    let itemText = collection.getItem(selectedKey)?.textValue ?? '';
+    let itemText = selectedKey ? collection.getItem(selectedKey)?.textValue ?? '' : '';
```
Comment (devongovett) [inline packages/@react-stately/combobox/src/useComboBoxState.ts:193]:
```suggestion
    let itemText = selectedKey != null ? collection.getItem(selectedKey)?.textValue ?? '' : '';
```

File: packages/@react-stately/combobox/src/useComboBoxState.ts [hunk 6]
```diff
@@ -176,13 +190,15 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateOptions<T
...
     setLastValue(itemText);
     setInputValue(itemText);
   };
 
   let lastSelectedKey = useRef(props.selectedKey ?? props.defaultSelectedKey ?? null);
-  let lastSelectedKeyText = useRef(collection.getItem(selectedKey)?.textValue ?? '');
+  let lastSelectedKeyText = useRef(
+    selectedKey ? collection.getItem(selectedKey)?.textValue ?? '' : ''
```
Comment (devongovett) [inline packages/@react-stately/combobox/src/useComboBoxState.ts:200]:
```suggestion
    selectedKey != null ? collection.getItem(selectedKey)?.textValue ?? '' : ''
```

(we need a lint for this somewhere :D)

File: packages/@react-stately/combobox/src/useComboBoxState.ts [hunk 7]
```diff
@@ -245,7 +261,7 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateOptions<T
     // This is to handle cases where a selectedKey is specified but the items aren't available (async loading) or the selected item's text value updates.
     // Only reset if the user isn't currently within the field so we don't erroneously modify user input.
     // If inputValue is controlled, it is the user's responsibility to update the inputValue when items change.
-    let selectedItemText = collection.getItem(selectedKey)?.textValue ?? '';
+    let selectedItemText = selectedKey ? collection.getItem(selectedKey)?.textValue ?? '' : '';
```
Comment (devongovett) [inline packages/@react-stately/combobox/src/useComboBoxState.ts:264]:
```suggestion
    let selectedItemText = selectedKey != null ? collection.getItem(selectedKey)?.textValue ?? '' : '';
```

File: packages/@react-stately/combobox/src/useComboBoxState.ts [hunk 8]
```diff
@@ -280,10 +296,10 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateOptions<T
   let commitSelection = () => {
     // If multiple things are controlled, call onSelectionChange
     if (props.selectedKey !== undefined && props.inputValue !== undefined) {
-      props.onSelectionChange(selectedKey);
+      props.onSelectionChange?.(selectedKey);
 
       // Stop menu from reopening from useEffect
-      let itemText = collection.getItem(selectedKey)?.textValue ?? '';
+      let itemText = selectedKey ? collection.getItem(selectedKey)?.textValue ?? '' : '';
```
Comment (devongovett) [inline packages/@react-stately/combobox/src/useComboBoxState.ts:302]:
```suggestion
      let itemText = selectedKey != null ? collection.getItem(selectedKey)?.textValue ?? '' : '';
```

File: packages/@react-stately/combobox/src/useComboBoxState.ts [hunk 9]
```diff
@@ -295,7 +311,7 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateOptions<T
 
   const commitValue = () => {
     if (allowsCustomValue) {
-      const itemText = collection.getItem(selectedKey)?.textValue ?? '';
+      const itemText = selectedKey ? collection.getItem(selectedKey)?.textValue ?? '' : '';
```
Comment (devongovett) [inline packages/@react-stately/combobox/src/useComboBoxState.ts:314]:
```suggestion
      const itemText = selectedKey != null ? collection.getItem(selectedKey)?.textValue ?? '' : '';
```

File: packages/@react-spectrum/tabs/src/Tabs.tsx [hunk 10]
```diff
@@ -55,7 +57,7 @@ interface TabsContext<T> {
   tabLineState: Array<DOMRect>
 }
 
-const TabContext = React.createContext<TabsContext<any>>(null);
+const TabContext = React.createContext<TabsContext<any> | {[key: string]: any}>({});
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-spectrum/tabs/src/Tabs.tsx]: it just required fewer changes was all, I've moved it back

File: packages/@react-spectrum/tabs/src/Tabs.tsx [hunk 11]
```diff
@@ -255,11 +257,12 @@ function TabLine(props: TabLineProps) {
  * The keys of the items within the <TabList> must match up with a corresponding item inside the <TabPanels>.
  */
 export function TabList<T>(props: SpectrumTabListProps<T>) {
-  const tabContext = useContext(TabContext);
+  const tabContext = useContext(TabContext)!;
   const {refs, tabState, tabProps, tabPanelProps} = tabContext;
   const {isQuiet, density, isEmphasized, orientation} = tabProps;
   const {selectedTab, collapsed, setTabListState} = tabState;
   const {tablistRef, wrapperRef} = refs;
+  console.log(tablistRef, wrapperRef)
```
Comment (devongovett) [inline packages/@react-spectrum/tabs/src/Tabs.tsx:265]:
```suggestion
```

**Review by devongovett — APPROVED** [review]

**Review by yihuiliao — APPROVED** [review]

**Pull request merged**